### PR TITLE
[FEATURE REQUEST] Manual removal of local storage

### PR DESCRIPTION
**Pull request by Aitorbp**
## Related Issues
App: https://github.com/owncloud/android/issues/4174

- [x] Add changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
- [x] Add feature to Release Notes in `ReleaseNotesViewModel.kt` creating a new `ReleaseNote()` with String resources (if required)

_____

## QA

checks performed: https://github.com/owncloud/android/pull/4334#issuecomment-2029357397

Reports:

- [X] (1) text overlaps icons https://github.com/owncloud/android/pull/4334#issuecomment-2029360123 [FIXED]
- [x] (2) Icon election https://github.com/owncloud/android/pull/4334#issuecomment-2029368667 [WONT FIX]

### CODE REVIEW

File: owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementViewModel.kt [hunk 1]
```diff
@@ -22,17 +23,39 @@ package com.owncloud.android.presentation.accounts
...
+
     fun getLoggedAccounts(): Array<Account> {
         return accountProvider.getLoggedAccounts()
     }
 
     fun getCurrentAccount(): Account? {
         return accountProvider.getCurrentOwnCloudAccount()
     }
+
+    fun cleanAccount(accountName: String) {
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementViewModel.kt:52]:
```suggestion
    fun cleanAccountLocalStorage(accountName: String) {
```

File: owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt [hunk 2]
```diff
@@ -196,6 +202,39 @@ class AccountsManagementActivity : FileActivity(), AccountsManagementAdapter.Acc
 
     }
 
+    override fun cleanAccount(account: Account) {
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt:205]:
```suggestion
    override fun cleanAccountLocalStorage(account: Account) {
```

File: owncloudApp/src/main/java/com/owncloud/android/usecases/files/RemoveFilesForAccountUseCase.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.owncloud.android.usecases.files
+
+import com.owncloud.android.domain.BaseUseCaseWithResult
+import com.owncloud.android.domain.files.FileRepository
+import com.owncloud.android.domain.files.usecases.RemoveFileUseCase
+
+class RemoveFilesForAccountUseCase(
```
Comment (JuancaG05) [inline owncloudApp/src/main/java/com/owncloud/android/usecases/files/RemoveFilesForAccountUseCase.kt:27]:
```suggestion
class RemoveLocalFilesForAccountUseCase(
```
This is important, otherwise it can be confused with a deletion (local and remote) of all the files of an account

File: owncloudApp/src/main/res/layout/account_item.xml [hunk 4]
```diff
@@ -80,6 +80,23 @@
...
+    <ImageView
+        android:id="@+id/cleanButton"
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:layout_gravity="center_vertical"
+        android:background="?android:attr/selectableItemBackground"
+        android:clickable="true"
+        android:paddingLeft="@dimen/standard_half_padding"
+        android:paddingTop="@dimen/standard_padding"
+        android:paddingRight="@dimen/standard_half_padding"
```
Comment (JuancaG05) [inline owncloudApp/src/main/res/layout/account_item.xml:92]: We shouldn't use `paddingLeft` or `paddingRight`, use `paddingStart` and `paddingEnd` instead

File: owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt [hunk 5]
```diff
@@ -538,6 +543,12 @@ interface FileDao {
             WHERE parentId = :folderId AND mimeType LIKE :mimeType || '%'
         """
 
+        private const val SELECT_FILES_FOR_ACCOUNT = """
+            SELECT *
+            FROM ${ProviderMeta.ProviderTableMeta.FILES_TABLE_NAME}
+            WHERE owner = :accountOwner AND keepInSync = '0'
+        """
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt:550]: You can improve this by getting just the files that we know are available locally... Imagine an account with 1 million files and we just want to remove 2. Does it make sense to return every file? It would be a waste of resources

File: owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt [hunk 6]
```diff
@@ -137,6 +137,11 @@ interface FileDao {
     @Query(SELECT_FILES_AVAILABLE_OFFLINE_FROM_EVERY_ACCOUNT)
     fun getFilesAvailableOfflineFromEveryAccount(): List<OCFileEntity>
 
+    @Query(SELECT_FILES_FOR_ACCOUNT)
+    fun getFilesForAccount(
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt:141]:
```suggestion
    fun getDownloadedFilesForAccount(
```

File: owncloudData/src/main/java/com/owncloud/android/data/files/datasources/LocalFileDataSource.kt [hunk 7]
```diff
@@ -44,6 +44,7 @@ interface LocalFileDataSource {
     fun getFilesWithSyncInfoAvailableOfflineFromAccountAsFlow(owner: String): Flow<List<OCFileWithSyncInfo>>
     fun getFilesAvailableOfflineFromAccount(owner: String): List<OCFile>
     fun getFilesAvailableOfflineFromEveryAccount(): List<OCFile>
+    fun getFilesForAccount(owner: String): List<OCFile>
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/files/datasources/LocalFileDataSource.kt:47]:
```suggestion
    fun getDownloadedFilesForAccount(owner: String): List<OCFile>
```

File: owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt:1]: ???
Please, don't commit unnecessary lines of code

File: owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt [hunk 8]
```diff
@@ -112,6 +112,17 @@ class OCLocalFileDataSourceTest {
         verify(exactly = 1) { fileDao.getFileByIdAsFlow(OC_FILE_ENTITY.id) }
     }
 
+    @Test
+    fun `getFilesForAccount returns a list of OCFile`() {
```
Comment (JuancaG05) [inline owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt:116]: You'll need to adapt the name of this test when changing the name of the method as well

File: owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt [hunk 9]
```diff
@@ -112,6 +112,17 @@ class OCLocalFileDataSourceTest {
...
+        every { fileDao.getFilesForAccount(OC_ACCOUNT_NAME) } returns listOf(OC_FILE_ENTITY)
+
+        val result = ocLocalFileDataSource.getFilesForAccount(OC_ACCOUNT_NAME)
+
+        assertEquals(listOf(OC_FILE), result)
+
+        verify(exactly = 1) { fileDao.getFilesForAccount(OC_ACCOUNT_NAME) }
+    }
```
Comment (JuancaG05) [inline owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt:124]: You have to add a test to see what happens when DAO returns an empty list. It's similar to `getFolderContent` method and its tests, didn't you see?

File: owncloudDomain/src/main/java/com/owncloud/android/domain/files/FileRepository.kt [hunk 10]
```diff
@@ -51,6 +51,7 @@ interface FileRepository {
     fun getFilesWithSyncInfoAvailableOfflineFromAccountAsFlow(owner: String): Flow<List<OCFileWithSyncInfo>>
     fun getFilesAvailableOfflineFromAccount(owner: String): List<OCFile>
     fun getFilesAvailableOfflineFromEveryAccount(): List<OCFile>
+    fun getFilesForAccount(owner: String): List<OCFile>
```
Comment (JuancaG05) [inline owncloudDomain/src/main/java/com/owncloud/android/domain/files/FileRepository.kt:54]:
```suggestion
    fun getDownloadedFilesForAccount(owner: String): List<OCFile>
```

File: owncloudData/src/test/java/com/owncloud/android/data/files/repository/OCFileRepositoryTest.kt [hunk 11]
```diff
@@ -198,6 +198,21 @@ class OCFileRepositoryTest {
...
+
+        val result = ocFileRepository.getFilesForAccount(OC_ACCOUNT_NAME)
+
+        assertEquals(listOf(OC_FILE), result)
+
+        verify(exactly = 1) {
+            localFileDataSource.getFilesForAccount(OC_ACCOUNT_NAME)
+        }
+    }
+
```
Comment (JuancaG05) [inline owncloudData/src/test/java/com/owncloud/android/data/files/repository/OCFileRepositoryTest.kt:215]: Tests for `OCFileRepository` are not refactored yet, but ok

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Several things to check here yet @Aitorbp

File: owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt [hunk 12]
```diff
@@ -196,6 +202,39 @@ class AccountsManagementActivity : FileActivity(), AccountsManagementAdapter.Acc
...
+        val dialog = AlertDialog.Builder(this)
+            .setTitle(getString(R.string.clean_data_account_title))
+            .setIcon(R.drawable.ic_warning)
+            .setMessage(getString(R.string.clean_data_account_message))
+            .setPositiveButton(getString(R.string.clean_data_account_button_yes)) { dialog, _ ->
+                accountsManagementViewModel.cleanAccount(account.name)
+                dialog.dismiss()
+            }
+            .setNegativeButton(R.string.drawer_close) { dialog, _ ->
```
[the remark this answers is not in the document]
Comment (jesmrec) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt:214]: `Cancel` means like an action referring the action describe above.

`Close` is not an action, is just dismissing the dialog without choosing any action.

Take the one that fits better. I'd get `Close`.

File: owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt [hunk 13]
```diff
@@ -87,6 +92,7 @@ class AccountsManagementActivity : FileActivity(), AccountsManagementAdapter.Acc
         account = accountsManagementViewModel.getCurrentAccount()
         onAccountSet(false)
 
+        subscribeToViewModels()
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt:95]: Any reason, better in `onCreate` yes

File: owncloudApp/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -789,4 +814,8 @@
     <string name="insecure_http_url_message_dialog">The URL you provided uses HTTP rather than the encrypted HTTPS protocol. If you continue, your communication will not be encrypted.</string>
     <string name="insecure_http_url_continue_button">Continue</string>
 
+    <string name="clean_data_account_message">All downloaded files that are not available offline will be cleaned up</string>
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/res/values/strings.xml:817]: Done

File: owncloudApp/src/main/res/values/strings.xml [hunk 15]
```diff
@@ -739,6 +739,31 @@
     <string name="release_notes_4_3_0_subtitle_5">The lens icon in the toolbar was removed when listing spaces in folder picker</string>
     <string name="release_notes_4_3_0_title_6">Add warning in http connections</string>
     <string name="release_notes_4_3_0_subtitle_6">Added a warning dialog in the login view when it is a http connection</string>
+    <string name="release_notes_4_3_0_title_8">New setting to removal manually local storage</string>
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/res/values/strings.xml:742]: Yeah, done

File: owncloudApp/src/main/res/values/strings.xml [hunk 16]
```diff
@@ -739,6 +739,31 @@
...
+    <string name="release_notes_4_2_0_title_7">Auto uploads to spaces</string>
+    <string name="release_notes_4_2_0_subtitle_7">Photos and videos from camera now saved inside spaces, not only in Personal</string>
+    <string name="release_notes_4_2_0_title_8">Manage password policy in live mode in Infinite Scale accounts</string>
+    <string name="release_notes_4_2_0_subtitle_8">Password policy for public links is handled in live mode with new items in the dialog</string>
+    <string name="release_notes_4_2_0_title_9">Video player improved</string>
+    <string name="release_notes_4_2_0_subtitle_9">A more advanced and optimized media playback experience for users, with improvements in performance and compatibility.</string>
+    <string name="release_notes_4_2_0_title_10">Prevent that two media files are playing at the same time</string>
+    <string name="release_notes_4_2_0_subtitle_10">The player handles the audio focus shifts, pausing one player if another starts</string>
+    <string name="release_notes_4_2_0_title_11">Minor bugfixes</string>
+    <string name="release_notes_4_2_0_subtitle_11">Some bugs were fixed to improve experience in the app</string>
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/res/values/strings.xml:766]: Yes... The rebase was a mess, I check the release notes for 4.3 but I didn´t notice about this. My mistake.

File: owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt (hunk 12 above)
[the remark this answers is not in the document]
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt:214]: No changes

File: owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementViewModel.kt (hunk 1 above)
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementViewModel.kt:52]: Done

File: owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt (hunk 2 above)
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/presentation/accounts/AccountsManagementActivity.kt:205]: Done

File: owncloudApp/src/main/java/com/owncloud/android/usecases/files/RemoveFilesForAccountUseCase.kt (hunk 3 above)
Comment (Aitorbp) [reply owncloudApp/src/main/java/com/owncloud/android/usecases/files/RemoveFilesForAccountUseCase.kt:27]: Done

File: owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt (hunk 5 above)
Comment (Aitorbp) [reply owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt:550]: You are right, done

File: owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt (hunk 6 above)
Comment (Aitorbp) [reply owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt:141]: Done

File: owncloudData/src/main/java/com/owncloud/android/data/files/datasources/LocalFileDataSource.kt (hunk 7 above)
Comment (Aitorbp) [reply owncloudData/src/main/java/com/owncloud/android/data/files/datasources/LocalFileDataSource.kt:47]: Done

File: owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt
Comment (Aitorbp) [reply owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt:1]: Done

File: owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt (hunk 8 above)
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt:116]: Adapted

File: owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt (hunk 9 above)
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt:124]: Yeah I was thinking to add it but in the function we don't have a condition to handle if the list is empty. I didn't find it very useful. Btw I added it, the rest of the tests have this implementation as well.

File: owncloudData/src/test/java/com/owncloud/android/data/files/repository/OCFileRepositoryTest.kt (hunk 11 above)
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/files/repository/OCFileRepositoryTest.kt:215]: Added the empty list test as well

File: changelog/unreleased/4334 [hunk 17]
```diff
@@ -0,0 +1,6 @@
+Enhancement: New setting to removal manually local storage
```
[the remark this answers is not in the document]
Comment (Aitorbp) [reply changelog/unreleased/4334]: Done

File: owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt [hunk 18]
```diff
@@ -538,6 +543,12 @@ interface FileDao {
             WHERE parentId = :folderId AND mimeType LIKE :mimeType || '%'
         """
 
+        private const val SELECT_FILES_FOR_ACCOUNT = """
```
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt:546]: Adapt the name as well:
```suggestion
        private const val SELECT_DOWNLOADED_FILES_FOR_ACCOUNT = """
```

File: owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt
Comment (JuancaG05) [inline owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt:1]: Why do you modify the header if this class has not been changed?

File: owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt (hunk 9 above)
Comment (JuancaG05) [reply owncloudData/src/test/java/com/owncloud/android/data/files/datasources/implementation/OCLocalFileDataSourceTest.kt:135]: We're giving a treatment for the result, in this case `toModel()`, we need to check this is working fine as well for empty lists

File: owncloudData/src/test/java/com/owncloud/android/data/files/repository/OCFileRepositoryTest.kt [hunk 19]
```diff
@@ -198,6 +199,32 @@ class OCFileRepositoryTest {
...
+    @Test
+    fun `getDownloadedFilesForAccount returns an empty list when datasource returns an empty list`() {
+        every { localFileDataSource.getDownloadedFilesForAccount(OC_ACCOUNT_NAME) } returns emptyList()
+
+        val result = ocFileRepository.getDownloadedFilesForAccount(OC_ACCOUNT_NAME)
+
+        assertEquals(emptyList<OCFile>(), result)
+
+        verify(exactly = 1) { localFileDataSource.getDownloadedFilesForAccount(OC_ACCOUNT_NAME) }
+    }
```
Comment (JuancaG05) [inline owncloudData/src/test/java/com/owncloud/android/data/files/repository/OCFileRepositoryTest.kt:226]: This is not needed, the result is what the datasource returns, we're not giving it a special treatment in-between

**Review by JuancaG05 — CHANGES_REQUESTED** [review]
Some more changes and ready to go @Aitorbp

File: owncloudData/src/test/java/com/owncloud/android/data/files/repository/OCFileRepositoryTest.kt (hunk 19 above)
Comment (Aitorbp) [reply owncloudData/src/test/java/com/owncloud/android/data/files/repository/OCFileRepositoryTest.kt:226]: Ok, deleted

File: owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt (hunk 18 above)
Comment (Aitorbp) [reply owncloudData/src/main/java/com/owncloud/android/data/files/db/FileDao.kt:546]: Done

File: owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt
Comment (Aitorbp) [reply owncloudData/src/main/java/com/owncloud/android/data/providers/LocalStorageProvider.kt:1]: Removed

**Review by JuancaG05 — APPROVED** [review]
LGTM 👍

**Comment by jesmrec** [comment]
Let's QA this one...

**Comment by jesmrec** [comment]
QA checks:

- [ ] UI changes
- [ ] Clean up oCIS account
- [ ] Clean up oC10 account
- [ ] Clean up one account out of several
- [ ] Resume uploads after cleaning up

**Comment by jesmrec** [comment]
### (1)

We have lost the layout adaption for long strings. Is this a regression?:

Screenshot 2024-04-01 at 10 06 03

Pixel 2, Android 11
`725873470`

**Comment by jesmrec** [comment]
### (2)

i don't want to look crazy, but the new icon looks like a brush more than a broom, not sure how clear it will be for users. This is how it looks like, any better idea @tbsbdr?:

Screenshot 2024-04-01 at 10 15 54

**Comment by tbsbdr** [comment]
as suggested by you: lets stick to the icon you suggested for now and show a dialog to that the user gets reassured within the text dialog, what will happen if he/she proceeds.

For the future, I can also think about using a "drawer menu" like you suggested and already made for files. This could be a better way to add more actions because we can fit more in it. In a drawer, we can also put in some text to explain what the icons mean.

**Comment by jesmrec** [comment]
So, the brush-broom icon will be the first approach. We'll work in future improvements
